Playwright: write new media upload test cases

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR adds new test cases not present in the older Selenium suite.

- new tests to check file upload functionality from Media gallery of various types (image, audio currently implemented).
- helper functions to generate test files.

#### Testing instructions

TeamCity
- ensure tests for Playwright Desktop E2E tests pass for this branch.

Local
- pull changes
- transpile code
- run `mocha --config .mocharc_playwright.yml` to verify the entire suite passes.

Depends on #53724

worldomonation committed:
Implement skeleton to generate test image, class methods to upload images and the test itself.

Refactor the media upload test to be parametrized.
Add new test case for invalid file.

Address code review comments.
- use Object.values for iteration when cleaning up.
- refactor the test to parametrize steps where valid file is being uploaded.
- combine the invalid file test into the main test suite.
- refactor the MediaPage.upload method to perform checks on the file instead of relying on other methods to do so.

Address Bart's PR comments:
- refactor test file generation methods to no longer take a parameter for the filename.
- move the cleanup action to the top level suite.
- change the fileChooser handling to click on buttons to interact with the file upload instead of setting the path programatically.

Re-enable test for simple site.
Revert the filechooser to the `setInputFile` method.

Remove dependency on sanitize-filename

WunderBart commented on test/e2e/specs/specs-playwright/wp-media__upload-spec.js:
```diff
@@ -0,0 +1,96 @@
+			} );
+
+			after( 'Clean up disk', async function () {
+				for ( const [ , filepath ] of Object.entries( testFiles ) ) {
```
```suggestion
				for ( const filepath of Object.values( testFiles ) ) {
```

worldomonation replied:
Thanks!

WunderBart commented on test/e2e/specs/specs-playwright/wp-media__upload-spec.js:
```diff
@@ -0,0 +1,96 @@
+
+			after( 'Clean up disk', async function () {
+				for ( const [ , filepath ] of Object.entries( testFiles ) ) {
+					MediaHelper.deleteFile( filepath );
```
Should this be `await`ed?

worldomonation replied:
`deleteFile` is synchronous, though I think it won't hurt to `await`.

worldomonation replied:
VSCodium tells me when I put `await` that it has no impact, so I've kept this section unchanged.

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -150,4 +158,45 @@ export class MediaPage extends BaseContainer {
+		// Each media item has the attribute `title=<filename.extension>` and Playwright
+		// is able to select elements by attribute.
+		await gallery.waitForSelector( `[title="${ basename }"]` );
+	}
```
I think we're mixing concerns here. It would be expected for the `upload` function to throw if the upload failed (as a self-validation, in case it's used for any other test), so in the test we'd just need to check if that error happened. What if the `upload` function was sth like this:

```js
async upload( filepath ) {
	await this.page.setInputFiles( selectors.addNewButton, filepath );

	const filename = path.basename( filepath );
	const itemSelector = `figure[title="${ filename }"]`;
	const result = await Promise.race( [
		this.page.waitForSelector( itemSelector ),
		this.page.waitForSelector( selectors.uploadRejectionNotice ),
	] );

	if ( ( await result.getAttribute( 'title' ) ) === filename ) {
		return result;
	}

	const noticeText = await result.$( '.notice__text' ).innerText();
	throw new Error( noticeText );
}
```

And then in the test:

```js
it( 'Upload a permitted file type and see it is added to the gallery' ) {
	const uploadedItem = await mediaPage.upload( filepath );
	assert( await uploadedItem.isVisible() );
}

it( 'Upload a forbidden file type and see the rejection notice' ) {
	try {
		await mediaPage.upload( filepath );
	} catch ( error ) {
		assert.match( error.message, /file type is not supported/ );
	}	
}
```

What do you think?

worldomonation replied:
I think you are totally right here. 

It's been 2 weeks since I wrote this and I am unsure why I decided to make methods for success and rejection. The wonders of spending some time away from the code.

I've revised the test and the MediaPage.upload() method along the lines of what you have suggested. For the spec itself, I have decided to parametrize the `it` step.

worldomonation commented:
I've moved this back to In Progress on the board due to needing to rebase and address PR comments.

WunderBart commented on test/e2e/specs/specs-playwright/wp-media__upload-spec.js:
```diff
@@ -0,0 +1,66 @@
+				try {
+					await mediaPage.upload( invalidFile );
+				} catch ( error ) {
+					console.log( 'message' + error.message );
```
Accidental log?

worldomonation replied:
Ah yes, oops.
As a side note, I go back and forth between "the `no-console` rule should be enabled!" and "oh no this rule is catching legitimate uses of `console.log`". 😭

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -19,6 +24,8 @@ const selectors = {
 	placeholder: '.is-placeholder',
 	editButton: 'button[data-e2e-button="edit"]',
+	addNewButton: '.media-library__upload-button-input',
+	uploadRejectionNotice: 'text=could not be uploaded because the file type is not supported',
```
We can be a bit less specific/brittle here
```suggestion
	uploadRejectionNotice: 'text=/could not be uploaded/i',
```

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -19,6 +24,8 @@ const selectors = {
 	items: '.media-library__list-item',
 	placeholder: '.is-placeholder',
 	editButton: 'button[data-e2e-button="edit"]',
+	addNewButton: '.media-library__upload-button-input',
```
Can we select based on the `Add new` text?

worldomonation replied:
Does not look like this can be done.

The `Add new` text is a `span` and is a sibling to the input. 
Since we are not clicking on the `Add new` button, selecting by the `Add new` text wouldn't work here:

```
 1) [WPCOM] Media: Upload: (desktop) @parallel
       Upload media files (Simple)
         Upload image and confirm addition to gallery:
     page.setInputFiles: Node is not an HTMLInputElement
=========================== logs ===========================
waiting for selector "span:text("Add new")"
  selector resolved to hidden <span data-e2e-sidebar="Add New" class="sidebar__menu-…>Add New</span>
```

WunderBart commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -140,4 +147,30 @@ export class MediaPage extends BaseContainer {
+			return result;
+		}
+
+		const noticeText = await this.page
+			.waitForSelector( '.notice__text' )
+			.then( ( element ) => element?.innerText() );
+		throw new Error( noticeText );
```
In this case the result will be the notice, so we should be able to go directly for its content:

```suggestion
		throw new Error( result.innerText() );
```

WunderBart commented on packages/calypso-e2e/src/media-helper.ts:
```diff
@@ -92,3 +94,62 @@ export function getFileName( {
+ * @param {string} filename Basename used for the generated test file.
+ * @returns {string} Full path on disk to the generated test file.
+ */
+export function getTestImage( filename: string ): string {
```
```suggestion
export function createTestImage( filename: string ): string {
```

WunderBart commented on packages/calypso-e2e/src/media-helper.ts:
```diff
@@ -92,3 +94,62 @@ export function getFileName( {
+ * @param {string} filename Basename used for the generated test file.
+ * @returns {string} Full path on disk to the generated test file.
+ */
+export function getTestAudio( filename: string ): string {
```
```suggestion
export function createTestAudio( filename: string ): string {
```

WunderBart commented on packages/calypso-e2e/src/media-helper.ts:
```diff
@@ -92,3 +94,62 @@ export function getFileName( {
 export function getDateString(): string {
```
```suggestion
export function createTimestamp(): string {
```

worldomonation replied:
I've made this change, but this also has knock-on changes in `BrowserManager` as well.

WunderBart commented on test/e2e/specs/specs-playwright/wp-media__upload-spec.js:
```diff
@@ -0,0 +1,66 @@
+import assert from 'assert';
+
+describe( DataHelper.createSuiteTitle( 'Media: Upload' ), function () {
+	const basename = MediaHelper.getDateString();
+	const testFiles = {
+		image: MediaHelper.getTestImage( basename + '.jpg' ),
+		audio: MediaHelper.getTestAudio( basename + '.mp3' ),
+	};
```
Since the basename is generated, can we move it to the create methods? Also because we shouldn't allow specifying the extension if [the source file's extension is fixed](https://github.com/Automattic/wp-calypso/pull/53828/files#diff-176fdd742302a8891bd16e0a7a9dbf6a9a51a70cd7c40a19ab7142d9bd3aca0eR144).

```suggestion
	const testFiles = {
		image: MediaHelper.getTestImage(),
		audio: MediaHelper.getTestAudio(),
	};
```

worldomonation replied:
The reason I get the `basename` in the test spec itself then pass that as the file name parameter to `create<file>` method is so they will all have the same name. It helped me with debugging, but I think what you say here is fair.

I've removed the parameter argument for `getTest<...>` calls and moved the basename generation into the `createTestFile` function itself.

worldomonation committed:
Rebase and update test to use Jest.

Use describe.each from Jest.

Remove commented lines

worldomonation committed:
Rebase, run linter and fix conflicts.

worldomonation committed:
Remove extraneous type dependency block.

worldomonation committed:
Run yarn deduplicate

worldomonation committed:
Restructure the `media-upload` spec to use Jest's `it.each` syntax.

worldomonation committed:
- remove errant console.log
- modify visibility check to wait for `visible`

dpasque commented on packages/calypso-e2e/src/lib/pages/media-page.ts:
```diff
@@ -12,6 +13,8 @@ const selectors = {
 	items: '.media-library__list-item',
 	placeholder: '.is-placeholder',
 	editButton: 'button[data-e2e-button="edit"]',
+	addNewButton: 'input[type="file"]',
```
A couple questions here:
1. Just to clarify - are we targeting the hidden input under the '.moxie-shim' div? I ask because this current selector seems a bit on the generic side, and am wondering if it's worth using something just a pinch more specific!
2. Because this isn't really the add new button, can we use a different property name? Something that suggests this is the selector for the hidden, test-only input field?

worldomonation replied:
1. I don't see a `.moxie-shim` div, but the gist is correct; the `upload` method will interact directly with the `input` element on the page. How about a selector like `input.media-library__upload-button-input`? 
2. Makes sense. Let's rename this to `fileInput` or the like?

dpasque replied:
Those changes both sound great!

worldomonation committed:
Change selector name for the upload button.

worldomonation committed:
- fix typo in test step name.
- if upload was not rejected, ensure the spinner is completely removed before proceeding.

dpasque commented:
New selectors for file look good, but it looks like there's an error occurring in the test script when doing the upload part of the test!

worldomonation committed:
Change the codepath for `upload` method.

worldomonation commented:
> New selectors for file look good, but it looks like there's an error occurring in the test script when doing the upload part of the test!

Yep, I was working on it and found the root cause.

This spec tests both Simple and Atomic sites, something the Selenium spec did not do.
Turns out Atomic and Simple sites have more difference than I expected.

The root cause of the failed tests were as follows:
- Simple: attempting to upload an unacceptable file results in near-immediate rejection.
- Atomic: uploading an unacceptable file is not rejected immediately; an entry in gallery appears briefly before the rejection notice is shown.

This behavior caused the previous iteration of the code to enter the 'Upload Successful' path for invalid files on Atomic even though it was not the case.

In my latest iteration, I have modified the `upload` method to wait for the upload to complete, then branch depending on if a rejection banner is shown.

dpasque commented:
Latest changes LGTM! You may have to rebase to fix the CI tests for Selenium - that failure (running super slow) is an artifact of the overnight Chrome version change
